apps/lockfile-explorer-web/src/linter.ts

```diff
@@ -0,0 +1,238 @@
+import { LockfileEntry, LockfileEntryFilter } from './parsing/LockfileEntry';
+
+interface ILockfileEntryGroup {
+  entryName: string;
+  versions: LockfileEntry[];
+}
+
+export const linter = (entries: LockfileEntry[]) => {
+  const packageEntries = entries.filter((entry) => entry.kind === LockfileEntryFilter.Package);
+  const projectEntries = entries.filter((entry) => entry.kind === LockfileEntryFilter.Project);
+
+  // Get the list of side-by-side versions
+  const reducedEntries = packageEntries.reduce((groups: { [key in string]: LockfileEntry[] }, item) => {
+    const group = groups[item.entryPackageName] || [];
+    group.push(item);
+    groups[item.entryPackageName] = group;
+    return groups;
+  }, {});
+  let groupedEntries: ILockfileEntryGroup[] = [];
+  for (const [packageName, entries] of Object.entries(reducedEntries)) {
+    groupedEntries.push({
+      entryName: packageName,
+      versions: entries
+    });
+  }
+
+  const sideBySide = groupedEntries.filter((entry) => entry.versions.length > 1);
+  const sideBySidePackageNames = new Set(sideBySide.map((s) => s.entryName));
```

Comment:
```suggestion
  const projectEntries: LockfileEntry[] = [];

  // Get the list of side-by-side versions
  const groupedPackageEntries: Map<string, LockfileEntry[]> = new Map();
  const sideBySidePackageNames: Set<string> = new Set();
  for (const entry of entries) {
    switch (entry.kind) {
      case LockfileEntryFilter.Package: {
        let groupedEntries: LockfileEntry[] | undefined = groupedPackageEntries.get(entry.entryPackageName);
        if (!groupedEntries) {
          groupedEntries = [];
          groupedPackageEntries.set(entry.entryPackageName, groupedEntries);
        }

        groupedEntries.push(entry);

        if (groupedEntries.length > 1) {
          sideBySidePackageNames.add(entry.entryPackageName);
        }
        break;
      }

      case LockfileEntryFilter.Project: {
        projectEntries.push(entry);
        break;
      }
    }
  }
```

Fewer loops and way fewer stack frames.